File: JeMPI_Apps/JeMPI_Configuration/src/main/scala/configuration/CustomLinkerProbabilistic.scala
package configuration

import java.io.{File, PrintWriter}
import scala.language.{existentials, postfixOps}


object CustomLinkerProbabilistic {

  private val classLocation = "../JeMPI_LibShared/src/main/java/org/jembi/jempi/linker"
  private val custom_className = "CustomLinkerProbabilistic"
  private val packageText = "org.jembi.jempi.linker"

  def parseRules(config: Config): Any = {
    val classFile: String = classLocation + File.separator + custom_className + ".java"
    println("Creating " + classFile)
    val file: File = new File(classFile)
    val writer: PrintWriter = new PrintWriter(file)

    val muList = for (
      t <- config.commonFields.filter(f => f.m.isDefined && f.u.isDefined)
    ) yield t

    writer.println(s"""package $packageText;""")
    if (muList.length == 0) {
      writer.println()
      writer.println(
        s"""
           |import org.jembi.jempi.shared.models.CustomMU;
           |import org.jembi.jempi.shared.models.CustomDemographicData;
           |
           |public final class $custom_className {
           |
           |  private $custom_className() {
           |  }
           |
           |  public static float probabilisticScore(final CustomDemographicData goldenRecord,
           |                                         final CustomDemographicData interaction) {
           |    return 0.0F;
           |  }
           |
           |  public static void updateMU(final CustomMU mu) {
           |  }
           |
           |  public static void checkUpdatedMU() {
           |  }
           |
           |  static CustomMU getMU() {
           |    return new CustomMU(null);
           |  }
           |
           |}""".stripMargin)
    } else {
      writer.println(
        s"""
           |import org.apache.commons.text.similarity.JaroWinklerSimilarity;
           |import org.jembi.jempi.shared.models.CustomDemographicData;
           |import org.jembi.jempi.shared.models.CustomMU;
           |
           |public final class $custom_className {
           |
           |   private static final JaroWinklerSimilarity JARO_WINKLER_SIMILARITY = new JaroWinklerSimilarity();
           |
           |   static Fields updatedFields = null;
           |
           |   private $custom_className() {
           |   }
           |""".stripMargin)

      writer.println("   static CustomMU getMU() {")
      writer.println("      return new CustomMU(")
      muList.zipWithIndex.foreach((mu, idx) => {
        writer.print(" " * 9 + s"LinkerProbabilistic.getProbability(currentFields.${Utils.snakeCaseToCamelCase(mu.fieldName)})")
        if (idx + 1 < muList.length)
          writer.println(",")
        else
          writer.println(
            """);
              |   }
              |""".stripMargin)
      })


      writer.println("   private record Fields(")
      muList.zipWithIndex.foreach((mu, idx) => {
        writer.print(s"""${" " * 9}LinkerProbabilistic.Field """)
        writer.print(Utils.snakeCaseToCamelCase(mu.fieldName))
        if (idx + 1 < muList.length)
          writer.println(",")
        else
          writer.println(
            s""") {
               |   }""".stripMargin)
          writer.println()
        end if
      })

      writer.println("   static Fields currentFields =")
      writer.print("      new Fields(")
      var margin = 0
      muList.zipWithIndex.foreach((field, idx) => {
        val m: Double = field.m.get
        val u: Double = field.u.get
        writer.print(" " * margin + s"new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, ${m}F, ${u}F)")
        if (idx + 1 < muList.length)
          writer.println(",")
          margin = 17
        else
          writer.println(");")
      })
      writer.println()
      writer.println(
        """   public static float probabilisticScore(
          |         final CustomDemographicData goldenRecord,
          |         final CustomDemographicData interaction) {
          |      // min, max, score, missingPenalty
          |      final float[] metrics = {0, 0, 0, 1.0F};""".stripMargin)
      muList.zipWithIndex.foreach((field, _) => {
        writer.println(" " * 6 + "LinkerProbabilistic.updateMetricsForStringField(metrics,")
        val fieldName = Utils.snakeCaseToCamelCase(field.fieldName)
        writer.println(" " * 54 + s"goldenRecord.$fieldName, interaction.$fieldName, currentFields" +
          s".$fieldName);")
      })
      writer.println(" " * 6 + "return ((metrics[2] - metrics[0]) / (metrics[1] - metrics[0])) * metrics[3];")
      writer.println(" " * 3 + "}")
      writer.println()
      writer.println("   public static void updateMU(final CustomMU mu) {")
      muList.zipWithIndex.foreach((field, idx) => {
        val fieldName = Utils.snakeCaseToCamelCase(field.fieldName)
        if (idx == 0)
          writer.print(" " * 6 + s"if (mu.$fieldName().m() > mu.$fieldName().u()")
        else
          writer.print(" " * 10 + s"&& mu.$fieldName().m() > mu.$fieldName().u()")
        end if
        if (idx + 1 < muList.length)
          writer.println()
        else
          writer.println(") {")
        end if
      })
      writer.println(" " * 9 + "updatedFields = new Fields(")
      muList.zipWithIndex.foreach((field, idx) => {
        val fieldName = Utils.snakeCaseToCamelCase(field.fieldName)
        writer.print(" " * 12 + s"new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, mu.$fieldName().m(), mu.$fieldName().u())")
        if (idx + 1 < muList.length)
          writer.println(",")
        else
          writer.println(");")
        end if

      })
      writer.println(" " * 6 + "}")
      writer.println("   }")
      writer.println()
      writer.println("}")
    }
    writer.flush()
    writer.close()
  }

}

File: JeMPI_Apps/JeMPI_Linker/src/main/java/org/jembi/jempi/linker/backend/CustomLinkerProbabilistic.java
package org.jembi.jempi.linker.backend;

import org.jembi.jempi.shared.models.CustomDemographicData;
import org.jembi.jempi.shared.models.CustomMU;

import java.util.Arrays;
import java.util.List;

import static org.jembi.jempi.linker.backend.LinkerProbabilistic.EXACT_SIMILARITY;
import static org.jembi.jempi.linker.backend.LinkerProbabilistic.JACCARD_SIMILARITY;
import static org.jembi.jempi.linker.backend.LinkerProbabilistic.JARO_SIMILARITY;
import static org.jembi.jempi.linker.backend.LinkerProbabilistic.JARO_WINKLER_SIMILARITY;

final class CustomLinkerProbabilistic {

   static final int METRIC_MIN = 0;
   static final int METRIC_MAX = 1;
   static final int METRIC_SCORE = 2;
   static final int METRIC_MISSING_PENALTY = 3;
   static final boolean PROBABILISTIC_DO_LINKING = false;
   static final boolean PROBABILISTIC_DO_VALIDATING = true;
   static final boolean PROBABILISTIC_DO_MATCHING = true;


   static ValidateFields updatedValidateFields = null;
   static MatchNotificationFields updatedMatchNotificationFields = null;

   private CustomLinkerProbabilistic() {
   }

   private record ValidateFields(
         LinkerProbabilistic.Field givenName,
         LinkerProbabilistic.Field familyName,
         LinkerProbabilistic.Field gender,
         LinkerProbabilistic.Field dob,
         LinkerProbabilistic.Field city,
         LinkerProbabilistic.Field phoneNumber) {
   }

   private record MatchNotificationFields(
         LinkerProbabilistic.Field givenName,
         LinkerProbabilistic.Field familyName,
         LinkerProbabilistic.Field gender,
         LinkerProbabilistic.Field dob,
         LinkerProbabilistic.Field city,
         LinkerProbabilistic.Field phoneNumber) {
   }

   static ValidateFields currentValidateFields =
      new ValidateFields(
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.8806329F, 0.0026558F),
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.9140443F, 6.275E-4F),
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.9468393F, 0.4436446F),
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.7856196F, 4.65E-5F),
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.8445694F, 0.0355741F),
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.84085F, 4.0E-7F));

   static MatchNotificationFields currentMatchNotificationFields =
      new MatchNotificationFields(
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.8806329F, 0.0026558F),
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.9140443F, 6.275E-4F),
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.9468393F, 0.4436446F),
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.7856196F, 4.65E-5F),
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.8445694F, 0.0355741F),
         new LinkerProbabilistic.Field(JARO_WINKLER_SIMILARITY, List.of(0.92F), 0.84085F, 4.0E-7F));

   static float linkProbabilisticScore(
         final CustomDemographicData goldenRecord,
         final CustomDemographicData interaction) {
      return 0.0F;
   }

   static float validateProbabilisticScore(
         final CustomDemographicData goldenRecord,
         final CustomDemographicData interaction) {

      // min, max, score, missingPenalty
      final float[] metrics = {0, 0, 0, 1.0F};
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.givenName, interaction.givenName, currentValidateFields.givenName);
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.familyName, interaction.familyName, currentValidateFields.familyName);
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.gender, interaction.gender, currentValidateFields.gender);
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.dob, interaction.dob, currentValidateFields.dob);
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.city, interaction.city, currentValidateFields.city);
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.phoneNumber, interaction.phoneNumber, currentValidateFields.phoneNumber);
      return ((metrics[METRIC_SCORE] - metrics[METRIC_MIN]) / (metrics[METRIC_MAX] - metrics[METRIC_MIN])) * metrics[METRIC_MISSING_PENALTY];
   }

   static float matchNotificationProbabilisticScore(
         final CustomDemographicData goldenRecord,
         final CustomDemographicData interaction) {

      // min, max, score, missingPenalty
      final float[] metrics = {0, 0, 0, 1.0F};
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.givenName, interaction.givenName, currentMatchNotificationFields.givenName);
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.familyName, interaction.familyName, currentMatchNotificationFields.familyName);
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.gender, interaction.gender, currentMatchNotificationFields.gender);
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.dob, interaction.dob, currentMatchNotificationFields.dob);
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.city, interaction.city, currentMatchNotificationFields.city);
      LinkerProbabilistic.updateMetricsForStringField(metrics,
                                                      goldenRecord.phoneNumber, interaction.phoneNumber, currentMatchNotificationFields.phoneNumber);
      return ((metrics[METRIC_SCORE] - metrics[METRIC_MIN]) / (metrics[METRIC_MAX] - metrics[METRIC_MIN])) * metrics[METRIC_MISSING_PENALTY];
   }
   public static void updateMU(final CustomMU mu) {
   }

}
